paasta_tools/check_spark_jobs.py

```diff
@@ -0,0 +1,170 @@
+        logger.warning(f'Unable to connect to {env_endpoint}: {e!r}')
+        return None
+
+    if response.status_code != 200:
+        logger.warning(f'Bad response from {env_endpoint}: {response.status_code}')
+        return None
+
+    try:
+        return response.json()['sparkProperties']
+    except (ValueError, KeyError) as e:
+        logger.warning(f'Unable to get sparkProperties for {framework.id}: {e!r}')
+        return None
+
+
+def guess_service(properties):
+    if not properties:
+        return None
+    for key, value in properties:
+        if key == 'spark.executorEnv.PAASTA_SERVICE':
+            service = value
+            break
+    else:
+        return None
+    if service.startswith(JUPYTER_PREFIX):
+        return service[len(JUPYTER_PREFIX):]
+    else:
+        return service
+
+
+def get_matching_framework_info(min_hours):
+    frameworks = mesos_tools.get_all_frameworks(active_only=True)
+    matching_info = []
+    min_timedelta = datetime.timedelta(hours=min_hours)
+    for framework in frameworks:
+        if not framework.active:
+            continue
+        if framework.get('principal') != 'spark':
+            continue
+        time_running = get_time_running(framework)
+        if time_running >= min_timedelta:
+            info = {
+                'id': framework.id,
+                'name': framework.name,
+                'webui_url': framework.get('webui_url'),
+                'service': guess_service(get_spark_properties(framework)),
+                'user': framework.user,
+                'time_running': str(time_running),
+            }
+            matching_info.append(info)
+
+    return matching_info
+
+
+def format_framework(info):
+    result = [f'{info["name"]} (running for {info["time_running"]})']
+    result.append(f'  user: {info["user"]}')
+    result.append(f'  job UI: {info["webui_url"]}')
+    return '\n'.join(result)
+
+
+def format_message_for_service(service, frameworks):
+    output = f'Found the following long-running Spark frameworks associated with service {service}.\n'
+    output += f'Please check why they are still running and terminate if appropriate.\n\n'
+    output += '\n'.join(format_framework(f) for f in frameworks)
+    return output
+
+
+def get_messages_by_service(results):
+    results_by_service = defaultdict(list)
+    for result in results:
+        service = result['service']
+        results_by_service[service].append(result)
+
+    return {
+        service: format_message_for_service(service, results)
+        for service, results in results_by_service.items()
+    }
+
+
+def notify_framework_owners(service, output):
+    overrides = {
+        'page': False,
+        'alert_after': 0,
+        'tip': 'Ask the user to check the job UI and terminate the job if appropriate.',
+        'runbook': 'http://y/spark-debug',
+    }
+    send_event(
+        service=service,
+        check_name=f'long_running_spark_jobs.{service}',
```

Comment:
This choice of check name is interesting.
On the plus side, we don't have to worry about some sort of "primary key" on the spark job (we will have plenty of adhoc things), so using service means we don't have to worry about lots of different unique sensu events.
On the minus side, if we have multiple long-running jobs per service we'll have grouping, and the check won't resolve until *all* the jobs are killed. They may cause confusion for users.

All this to say: I support your choice of check name

Reply:
Thanks for the support :) I think it's ok for the check not to resolve until everything is killed, since the text says the alert is for a particular service and lists all the frameworks.

That reminds me, do I need to send OK events for all services that **don't** have long-running jobs? As far as I can tell, if I don't, these alerts won't resolve.

Reply:
Haha! Yes, you need to send OK events. (I forgot too)